test/framework/src/main/java/org/elasticsearch/transport/nio/channel/TcpReadContext.java

```diff
@@ -0,0 +1,199 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.transport.nio.channel;
+
+import org.elasticsearch.common.bytes.BytesArray;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.elasticsearch.monitor.jvm.JvmInfo;
+import org.elasticsearch.transport.TcpHeader;
+import org.elasticsearch.transport.TcpTransport;
+import org.elasticsearch.transport.nio.TcpReadHandler;
+
+import java.io.IOException;
+import java.io.StreamCorruptedException;
+import java.nio.ByteBuffer;
+
+public class TcpReadContext implements ReadContext {
+
+    private static final int READ_INCOMPLETE = 0;
+    private static final int PING = 1;
+    private static final int READ_COMPLETE = 2;
+    private static final int PEER_CLOSED = 3;
+    private static final long NINETY_PER_HEAP_SIZE = (long) (JvmInfo.jvmInfo().getMem().getHeapMax().getBytes() * 0.9);
+
+    private final TcpReadHandler handler;
+    private final NioSocketChannel channel;
+    private final ByteBuffer headerBuffer = ByteBuffer.allocate(TcpHeader.MARKER_BYTES_SIZE + TcpHeader.MESSAGE_LENGTH_SIZE);
+
+    private int bytesRead = 0;
+    private int headerBytesRemaining = 6;
+    private int messageBytesLength = -1;
+    private int messageBytesRemaining = -1;
+    private ByteBuffer messageBuffer;
+
+    public TcpReadContext(NioSocketChannel channel, TcpReadHandler handler) {
+        this.handler = handler;
+        this.channel = channel;
+    }
+
+    @Override
+    public int read() throws IOException {
+        bytesRead = 0;
+//        BytesReference ref = new PagedBytesReference(selector.getBigArrays(),
```

Comment:
This is where I was playing around with using our byte array recycler for the reads. I decided anything like that was not necessary for this PR. I'll remove the comments.